coa/pkg/apis/v1alpha2/providers/states/redisstate/redisstate.go

```diff
@@ -0,0 +1,378 @@
+		if err != nil {
+			return entry.Value.ID, v1alpha2.NewCOAError(nil, fmt.Sprintf("old redis state %s status cannot be parsed", entry.Value.ID), v1alpha2.InternalError)
+		}
+		var newStatusDict map[string]interface{}
+		_, newStatusDict, err = getStatusDictFromMarshalStateEntryBody(body)
+		if err != nil {
+			return entry.Value.ID, v1alpha2.NewCOAError(nil, fmt.Sprintf("new redis state %s cannot be parsed", entry.Value.ID), v1alpha2.InternalError)
+		}
+		for k, v := range newStatusDict {
+			oldStatusDict[k] = v
+		}
+		oldEntryDict["status"] = oldStatusDict
+		body, _ = json.Marshal(oldEntryDict)
+		_, err = r.Client.HSet(key, "values", string(body)).Result()
+		return entry.Value.ID, err
+	}
+
+	properties := map[string]interface{}{
+		"values": string(body),
+		"etag":   entry.Value.ETag,
+	}
+	_, err = r.Client.HSet(key, properties).Result()
+	return entry.Value.ID, err
+}
+
+func (r *RedisStateProvider) List(ctx context.Context, request states.ListRequest) ([]states.StateEntry, string, error) {
+	_, span := observability.StartSpan("Redis State Provider", ctx, &map[string]string{
+		"method": "List",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+
+	var entities []states.StateEntry
+	var keyPrefix string
+	keyPrefix, err = getObjectTypePrefixForList(request.Metadata)
+	if err != nil {
+		rLog.Errorf("  P (Redis State): list states failed to get key prefix with error %s, traceId: %s", err.Error(), span.SpanContext().TraceID().String())
+		return entities, "", err
+	}
+	if n, ok := request.Metadata["namespace"]; ok {
+		if nstring, ok := n.(string); ok && nstring != "" {
+			keyPrefix = keyPrefix + separator + nstring
+		}
+	}
+	rLog.Debugf("  P (Redis State): list states with keyPrefix %s, traceId: %s", keyPrefix, span.SpanContext().TraceID().String())
+
+	filter := fmt.Sprintf("%s%s*", keyPrefix, separator)
+	var cursor uint64 = 0
+	var keys []string
+
+	for {
+		var err error
+		keys, cursor, err = r.Client.Scan(cursor, filter, entryCountPerList).Result()
+		if err != nil {
+			rLog.Errorf("  P (Redis State): failed to get all the keys matching pattern %s: %+v", keyPrefix, err)
+		}
+
+		for _, key := range keys {
+			result, err := r.Client.HGetAll(key).Result()
+			if err != nil || len(result) == 0 {
+				rLog.Errorf("  P (Redis State): failed to get entry for key %s: %+v", key, err)
+				continue
+			}
+			parts := strings.Split(key, separator)
+			if len(parts) != 3 {
+				rLog.Errorf("  P (Redis State): key is not valid %s: %+v", key, err)
+				continue
+			}
+			entry, err := CastRedisPropertiesToStateEntry(parts[2], result)
+			if err != nil {
+				rLog.Errorf("  P (Redis State): failed to cast entry for key %s: %+v", key, err)
+				continue
+			}
+			if request.FilterType != "" && request.FilterValue != "" {
+				var match bool
+				match, err = states.MatchFilter(entry, request.FilterType, request.FilterValue)
```

Comment:
I can understand your concern that RediSearch is under (i) Redis Source Available License 2.0 (RSALv2) or (ii) the Server Side Public License v1 (SSPLv1), not all Symphony deployment can leverage RediSearch. Under this case, we can only filter data property in client side. But there's still some follow up items we need to consider in future,
1) Using client side filter, how can we support paging if needed in future?
2) If Symphony is deployed with standalone and users want to supply a redis server with RediSearch loaded, we can extend redis state provider to leverage this feature.

Considering the need of above two follow up tasks are not in vision, I am OK we can firstly handle the redis state filter in memory to unblock the crash consistency issue which has higher priority.

Reply:
Issue has been opened per offline discussion.